cid-decode/src/main.rs

```diff
@@ -1,103 +1,209 @@
 #[macro_use]
 extern crate log;
+#[macro_use]
+extern crate lazy_static;
+
+mod multicodecs;
 
+use failure::err_msg;
 use ipfs_resolver_common::{logging, Result};
+use multicodecs::Tag;
 use std::collections::HashMap;
 use std::convert::TryFrom;
+use std::io;
 use std::io::{BufRead, BufReader};
-use std::{io, panic};
 
 fn main() -> Result<()> {
     dotenv::dotenv().ok();
     logging::set_up_logging()?;
 
-    let mut rdr = BufReader::new(io::stdin());
-    let mut s = String::new();
+    group_and_count_cid_by_metadata(&mut io::stdin(), &mut io::stdout(), &mut io::stderr())
+}
 
+fn group_and_count_cid_by_metadata(
+    input: &mut impl io::Read,
+    mut output: &mut impl io::Write,
+    mut error_output: &mut impl io::Write,
+) -> Result<()> {
+    let mut rdr = BufReader::new(input);
+    let mut buffer = String::new();
     let mut results: HashMap<_, usize> = HashMap::new();
 
-    let h = panic::take_hook();
-
-    // Set panic hook to suppress warnings (bruh.)
-    panic::set_hook(Box::new(|_info| {
-        // do nothing
-    }));
-    while let Ok(n) = rdr.read_line(&mut s) {
+    let mut line_no = 0;
+    while let Ok(n) = rdr.read_line(&mut buffer) {
         if n == 0 {
-            // Restore panic hook.
-            panic::set_hook(h);
+            //EOF
             results
                 .into_iter()
-                .for_each(|(k, v)| println!("{},{}", k, v));
+                .try_for_each(|(k, v)| writeln!(&mut output, "{},{}", k, v))?;
             return Ok(());
         }
+        line_no += 1;
+        debug!("working on row {}: ({})", line_no, buffer.trim());
 
-        debug!("working on {}", s.trim());
-
-        let res = match do_single(s.trim()) {
-            Err(_) => "invalid".to_string(),
+        let res = match do_single(buffer.trim()) {
             Ok(m) => format!(
-                "{:?}:{:?}:{:?}:{}:{}",
-                m.base,
-                m.version,
-                m.codec,
-                if let Some(h) = m.hash {
-                    format!("{:?}", h)
-                } else {
-                    "invalid".to_string()
-                },
-                m.hash_len
+                "{:?}:{:?}:{}:{}:{}",
+                m.base, m.version, m.codec, m.hash, m.hash_len
             ),
+            Err(err) => {
+                // If we get an error while parsing the input we log a warning to stderr and ignore the corresponding line
+                writeln!(
+                    &mut error_output,
+                    "Ignoring line {} because of error {} (Line: {}...)",
```

Comment:
You should debug-print the error using `{:?}`, which will also print the backtrace.
I'd reformat this as
`"Ignoring line {} ({:10}...) because of error: {:?}", line_no, buffer.trim(), err`

Reply:
I did this slightly different but with the same effect.